### PR TITLE
docs: Add new 'Animating widgets' how-to guide.

### PR DESCRIPTION
**Pull request by qchong**
Staged preview:
- [Animating widgets](https://flutter-io-staging-69a8a.firebaseapp.com/animating-widgets/) how-to guide (new)
- [Animations in Flutter](https://flutter-io-staging-69a8a.firebaseapp.com/animations/) developer guide (updated)

@abarth @HansMuller ptal.

### CODE REVIEW

File: _data/sidebars/home_sidebar.yml [hunk 1]
```diff
@@ -98,6 +98,10 @@ entries:
       url: /text-input/
       output: web
 
+    - title: Animate Widgets
+      url: /animating-widgets/
```
Comment (abarth) [inline _data/sidebars/home_sidebar.yml]: Probably the title and URL should match.

File: animating-widgets.md [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+
+## Introduction
+
+You can use animations to make the user experience on your mobile app feel more
+intuitive, dynamic, and polished. Animations are especially useful when the
+screen changes state, such as when content loads or when new actions become
+available.
+
+The [`flutter/animation`](https://docs.flutter.io/flutter/animation/animation-library.html)
+package in the Flutter SDK provides APIs to create animation effects in user
```
Comment (abarth) [inline animating-widgets.md]: Technically `animation` is a library rather than a package.

File: animating-widgets.md [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+following:
+
+1. Attach an animation object to a widget or a listener to detect changes to the
+animation object.
+2. Specify how the the animation is *interpolated* (that is, how the animation
+objects’s value and status changes over
+time) as it runs.
+
+Based on the animation object’s properties at a specific frame, Flutter
+dynamically modifies your widget visual appearence, and rebuilds the widget
```
Comment (abarth) [inline animating-widgets.md]: widget's visual appearance

File: animating-widgets.md [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+Flutter's flexible animation model allows you to define interpolations for
+spatial positioning, color gradients, image scaling, and others. You can define
+your own subclass of
+[`Animation`](https://docs.flutter.io/flutter/animation/Animation-class.html)
+or use an animation subclass that the framework provides.
+
+To learn more about
+using the other key classes in the
+[`flutter/animation`](https://docs.flutter.io/flutter/animation/animation-library.html)
+package, see the [Animations in Flutter](/animations) guide.
```
Comment (abarth) [inline animating-widgets.md]: s/package/library/

File: animating-widgets.md [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+
+## Render a basic widget animation
+
+To render your widget with an animation, set one or more of the widget’s
+members to the animation object, then use the animation’s value to decide how
+to draw your object.
```
Comment (abarth) [inline animating-widgets.md]: draw is kind of an unusual case.  It's much more common to use the animation's value to decide how to build child widgets.

File: animating-widgets.md [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+
+If you are animating a stateful widget, you can notify your widget of changes
+to the animation’s values by using the
+[`addListener()`](https://docs.flutter.io/flutter/animation/Animation/addListener.html)
+method. The framework invokes the listener callback whenever the value of the
+animation changes. Typically, you’ll want to call the
+[`setState()`](http://docs.flutter.io/flutter/material/State/setState.html)
+method when the animation value change occurs, to trigger the framework to
+rebuild the widget tree.
```
Comment (abarth) [inline animating-widgets.md]: Rather than doing this, we might want to recommend that people use `AnimatedWidget` or `AnimatedBuilder`.

File: animating-widgets.md [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+
+class LogoApp extends StatefulWidget {
+  LogoAppState createState() => new LogoAppState();
+}
+
+class LogoAppState extends State<LogoApp> {
+  Animation<double> animation;
+
+  initState() {
+    AnimationController controller = new AnimationController(
```
Comment (abarth) [inline animating-widgets.md]: You need the `vsync` parameter now.

File: animating-widgets.md [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+      duration: const Duration(milliseconds: 5000)
+    );
+    animation = new Tween(begin: 0.0, end: 200.0).animate(controller);
+    animation.addListener(() {
+      setState(() {
+      // the state that has changed here is the animation object’s value
+      });
+    });
+    controller.forward();
+    super.initState();
```
Comment (abarth) [inline animating-widgets.md]: The typical style is to call `super.initState` first.

File: animating-widgets.md [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+  }
+
+  Widget build(BuildContext context) {
+    return new Center(
```
Comment (abarth) [inline animating-widgets.md]: This is a good application of `AnimatedBuilder`.  That will let you skip the `animation` member variable.  You can use a build function like:

```dart
Widget build(BuildContext context) {
  Animation<double> animation = new Tween(begin: 0.0, end: 200.0).animate(_controller);
  return new Container(
    padding: const EdgeInsets.symmetric(vertical: 10.0),
    child: new AnimatedBuilder(
      animation: animation,
      builder: (BuildContext context) {
        return new Image.asset(
          name: 'assets/flutter_logo.png',
          width: animation.value,
          height: animation.value,
          fit: ImageFit.contain
        );
      }
  )
}
```

File: animating-widgets.md [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+    AnimationController controller = new AnimationController(
+      duration: const Duration(milliseconds: 5000)
+    );
+    animation = new Tween(begin: 0.0, end: 200.0).animate(controller);
+    animation.addListener(() {
+      setState(() {
+      // the state that has changed here is the animation object’s value
+      });
+    });
+    controller.forward();
```
Comment (abarth) [inline animating-widgets.md]: You also need to hold onto the controller in a member variable and call it's `dispose` method in the State's `dispose` method.  If you don't do this, the animation will continue to run if your widget is removed from the tree, which is a waste of resources and will trigger an asset.

File: animating-widgets.md [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+}
+
+void main() {
+  runApp(
+    new MaterialApp(
+      title: 'My Animation Test',
+      routes: <String, WidgetBuilder> {
+        '/': (BuildContext context) => new LogoApp()
+      }
+    )
```
Comment (abarth) [inline animating-widgets.md]: You can simplify this as:

```dart
runApp(
  new MaterialApp(
    title: 'My Animation Test',
    home: new LogoApp()
  )
);
```

File: animating-widgets.md [hunk 12]
```diff
@@ -0,0 +1,293 @@
...
+  LogoApp({Key key, Animation<double> animation})
+      : super(key: key, animation: animation);
+
+  Widget build(BuildContext context) {
+    return new Center(
+      child: new Container(
+        margin: new EdgeInsets.symmetric(vertical: 10.0),
+        height: animation.value,
+        width: animation.value,
+        child: new AssetImage(
```
Comment (abarth) [inline animating-widgets.md]: This should be `Image.asset`.  Also, the width and height parameter should move to the image (it's more efficient).  Finally, width should be listed before height.

File: animating-widgets.md [hunk 13]
```diff
@@ -0,0 +1,293 @@
...
+}
+
+void main() {
+  final AnimationController controller = new AnimationController(
+    duration: const Duration(milliseconds: 5000)
+  );
+  final animation = new Tween(
+    begin: 0.0,
+    end: 200.0)
+  .animate(controller);
```
Comment (abarth) [inline animating-widgets.md]: I'm not sure this is a good example.  It's extremely unusual to create an AnimationController inside the `main` function.  Also, now with the `vsync` parameter, you'll have trouble figuring out which `vsync` you want to attach your animation to.

I'm not sure it's worth presenting these two examples.  Instead, we should just use `AnimatedBuilder` in the first example.  That's the code you'd normally write and it's simple enough to explain in one go.

File: animating-widgets.md [hunk 14]
```diff
@@ -0,0 +1,293 @@
...
+      }
+    )
+  );
+  controller.forward();
+}
+```
+
+For more complex animations involving additional state, consider using the
+[`AnimatedBuilder`](http://docs.flutter.io/flutter/material/AnimatedBuilder-class.html)
+class.
```
Comment (abarth) [inline animating-widgets.md]: I'd skip this sentence.  `AnimatedBuilder` is good for simple animations too.

File: animating-widgets.md [hunk 15]
```diff
@@ -0,0 +1,293 @@
...
+
+```dart
+import 'package:flutter/material.dart';
+import 'package:flutter/animation.dart';
+
+class LogoApp extends AnimatedWidget {
+
+  // Declare the Tween objects as static because we don't change them.
+  static final _opacityTween = new Tween<double>(begin: 1.0, end: 0.3);
+  static final _sizeTween = new Tween<double>(begin: 0.0, end: 300.0);
```
Comment (abarth) [inline animating-widgets.md]: We would normally declare these at the library scope.

File: animating-widgets.md [hunk 16]
```diff
@@ -0,0 +1,293 @@
...
+      : super(key: key, animation: animation);
+
+  Widget build(BuildContext context) {
+    return new Center(
+      child: new Opacity(
+        opacity: _opacityTween.evaluate(animation),
+        child: new Container(
+          margin: new EdgeInsets.symmetric(vertical: 10.0),
+          height: _sizeTween.evaluate(animation),
+          width: _sizeTween.evaluate(animation),
```
Comment (abarth) [inline animating-widgets.md]: Better to evaluate once and re-use the value than to call `evaluate` twice.

File: animating-widgets.md [hunk 17]
```diff
@@ -0,0 +1,293 @@
...
+        )
+      )
+    );
+  }
+}
+
+void main() {
+  final AnimationController controller = new AnimationController(
+    duration: const Duration(milliseconds: 5000)
+  );
```
Comment (abarth) [inline animating-widgets.md]: Again, this pattern of creating an AnimationController is `main` is extremely unusual and isn't a practice we should recommend.  Instead, we should host the AnimationController in a State object so that it can get the proper `vsync` signal and participate in the rest of the lifecycle.

File: animating-widgets.md [hunk 18]
```diff
@@ -0,0 +1,293 @@
...
+    )
+  );
+  controller.forward();
+}
+```
+
+## Dispose your widget animation
+
+It’s good practice to dispose your animation controller and free up your
+resources when they are no longer needed. To do so, call the
```
Comment (abarth) [inline animating-widgets.md]: It's more than a good practice.  It's required for your code to be correct and you'll now get an assert if you don't do it.

File: animating-widgets.md [hunk 19]
```diff
@@ -0,0 +1,293 @@
...
+```dart
+class LogoAppState extends State<LogoApp> {
+
+  // Implementation code omitted for brevity.
+
+  @override
+  void dispose() {
+     controller.dispose();
+     super.dispose();
+  }
```
Comment (abarth) [inline animating-widgets.md]: We should be doing this for all the examples throughout.  It's not really optional.

File: animations.md [hunk 20]
```diff
@@ -17,27 +17,34 @@ functions directly by reading their current value and listening to their
...
+notifies all the listeners added with
 [`addListener`](http://docs.flutter.io/flutter/animation/Animation/addListener.html).
 Typically, a [`State`](http://docs.flutter.io/flutter/widgets/State-class.html)
 object that listens to an animation will call
 [`setState`](http://docs.flutter.io/flutter/widgets/State/setState.html) on
 itself in its listener callback to notify the widget system that it needs to
 rebuild with the new value of the animation.
 
-This pattern is so common that there are two widgets that help widgets rebuild
+     This pattern is so common that there are two widgets that help widgets rebuild
```
Comment (abarth) [inline animations.md]: spurious indent?

File: animations.md [hunk 21]
```diff
@@ -63,20 +69,35 @@ to 1.0 will be `dismissed` when their value is 0.0. An animation might then run
...
+[`AnimationController`](http://docs.flutter.io/flutter/animation/AnimationController-class.html)
+class lets you define important characteristics of the animation, such as its
+duration and playback direction (forward or reverse).
+
+The following snippet shows how you might create an animation that runs for
+500 milliseconds:
+
+```dart
+final AnimationController controller = new AnimationController(
+  duration: const Duration(milliseconds: 500)
```
Comment (abarth) [inline animations.md]: We'll need to update this with `vsync: this`

File: animations.md [hunk 22]
```diff
@@ -85,7 +106,90 @@ from 1.0 to 0.0). Similarly, you can create a
...
+and others.
+
+The following snippet shows how you might create a curved animation object that
+wraps a linear animation produced by an animation controller and applies the
+[`easeIn`](http://docs.flutter.io/flutter/animation/Curves/easeIn-constant.html)
+curve type to it.
+
+```dart
+final AnimationController controller =
+    new AnimationController(duration: const Duration(milliseconds: 500));
```
Comment (abarth) [inline animations.md]: We usually use two-space indent.

File: animations.md [hunk 23]
```diff
@@ -85,7 +106,90 @@ from 1.0 to 0.0). Similarly, you can create a
...
+
+final curve = new CurvedAnimation(
```
Comment (abarth) [inline animations.md]: I'd skip the `final` and supply a type instead.

File: animations.md [hunk 24]
```diff
@@ -85,7 +106,90 @@ from 1.0 to 0.0). Similarly, you can create a
...
+interpolates between the double values of -200.0 to 0.0.
+
+```dart
+final Tween doubleTween = new Tween<double>(begin: -200.0, end: 0.0);
+```
+
+You can also specify a progression between two colors. For example:
+
+```dart
+final Tween colorTween =
```
Comment (abarth) [inline animations.md]: I'd drop the `final` from all these example.  Here I'd use the more specific ColorTween type.

File: animations.md [hunk 25]
```diff
@@ -85,7 +106,90 @@ from 1.0 to 0.0). Similarly, you can create a
...
++ [`ThemeDataTween`](https://docs.flutter.io/flutter/material/ThemeDataTween-class.html):
+Interpolation between
+[`ThemeData`](https://docs.flutter.io/flutter/material/ThemeData-class.html) objects.
+
+The following snippet shows how you might create a
+[`Tween`](http://docs.flutter.io/flutter/material/Tween-class.html) object that
+interpolates between the double values of -200.0 to 0.0.
+
+```dart
+final Tween doubleTween = new Tween<double>(begin: -200.0, end: 0.0);
```
Comment (abarth) [inline animations.md]: Dropping the type parameter here is dodgy.  Better to keep it in.

File: animations.md [hunk 26]
```diff
@@ -85,7 +106,90 @@ from 1.0 to 0.0). Similarly, you can create a
...
+```
+
+You can also specify a progression between two colors. For example:
+
+```dart
+final Tween colorTween =
+    new ColorTween(begin: Colors.transparent, end: Colors.black54);
+```
+
+To apply a Tween to a given animation object, use the [`animate()`]()
```
Comment (abarth) [inline animations.md]: Extra `()`

File: animations.md [hunk 27]
```diff
@@ -85,7 +106,90 @@ from 1.0 to 0.0). Similarly, you can create a
...
+snippet shows how you might create an animation represented by integral values
+from 0 to 255, over a duration of 500 milliseconds, using linear interpolation.
+
+```dart
+final AnimationController animation =
+    new AnimationController(duration: const Duration(milliseconds: 500));
+
+Animation<int> alpha = new IntTween(
+  begin: 0,
+  end: 255)
```
Comment (abarth) [inline animations.md]: This whitespace is odd.  I'd just put it all on one line.

File: animations.md [hunk 28]
```diff
@@ -104,26 +208,24 @@ function.
...
+
+Use these methods if you need to combine a tween with an animation to get a
+concrete value:
+
++ [`evaluate()`](http://docs.flutter.io/flutter/animation/Tween/evaluate.html):
+Evaluates the tween at the current value of an animation. This approach is most
+useful for widgets that are already listening to the animation and hence
+rebuilding whenever the animation changes value.
++ [`animate()`](http://docs.flutter.io/flutter/animation/Animatable/animate.html)
+   Animates the tween based on the animation object's properties. Rather than
```
Comment (abarth) [inline animations.md]: Inconsistent indent between these two bullets.

File: animating-widgets.md (hunk 13 above)
Comment (jimbeveridge) [inline animating-widgets.md]: I strongly disagree that it's "simple enough". I originally wrote my animation tutorial because of the amount of effort it took to understand how AnimatedBuilder worked. Unfortunately the AnimatedBuilder example was removed from this document, as was all of the discussion about how AnimatedBuilder worked.

File: animating-widgets.md (hunk 14 above)
Comment (jimbeveridge) [inline animating-widgets.md]: I disagree with this comment for the intended audience of this document. The sentence is to help direct people with their progression of learning.

File: animating-widgets.md (hunk 11 above)
Comment (jimbeveridge) [inline animating-widgets.md]: Why were all of the "new MaterialApp" statements added to the examples? These are not Material Apps and so similar statements were removed in early code reviews of these examples.

File: animating-widgets.md [hunk 29]
```diff
@@ -0,0 +1,293 @@
...
+      : super(key: key, animation: animation);
+
+  Widget build(BuildContext context) {
+    return new Center(
+      child: new Container(
+        margin: new EdgeInsets.symmetric(vertical: 10.0),
+        height: animation.value,
+        width: animation.value,
+        child: new AssetImage(
+           name: 'assets/flutter_logo.png',
```
Comment (jimbeveridge) [inline animating-widgets.md]: Why were these changed from web-based images to local images? With web-based images, readers can copy and paste the examples and they'll work. With local examples, readers have to create a project with the directory structure exactly right, and get the flutter.yaml file right, none of which is shown here. This was identified during usability testing of the original examples.

File: animating-widgets.md (hunk 13 above)
Comment (jimbeveridge) [inline animating-widgets.md]: The point of this document is to teach people how the animations work, not to provide boilerplate to copy and paste. I sounds like the document should discourage readers from using the low level calls, which is fine, but (IMO) understanding how it works under the covers was key to building up to how AnimatedBuilder works.

Comment (abarth) [inline animating-widgets.md]: That makes sense.  However, we should be aware that people are going to copy-and-paste code samples.  If we're going to show people code that we don't want them to copy (e.g., missing dispose calls and starting AnimationControllers in main), then we should make it more obvious that these are code snippets and not complete executable samples.

File: animating-widgets.md (hunk 11 above)
Comment (abarth) [inline animating-widgets.md]: Yeah, we can skip that and just run `LogoApp` directly.

File: animating-widgets.md (hunk 13 above)
Comment (jimbeveridge) [inline animating-widgets.md]: This code was written last January and things have changed since then, so, instead of making them snippets, I think we should do your proposed changes to add dispose() calls and take the AnimationControllers out of main(). Even if we are showing low level calls, we should still show the proper way to use them.

File: animating-widgets.md (hunk 19 above)
Comment (jimbeveridge) [inline animating-widgets.md]: +1

File: animations.md (hunk 21 above)
Comment (jimbeveridge) [inline animations.md]: And we need to add some expository text about vsync.

File: _data/sidebars/home_sidebar.yml (hunk 1 above)
Comment (qchong) [inline _data/sidebars/home_sidebar.yml]: The titles in the sidebar (left-nav) are shortened by design, to make them easier to scan. The full title ('Animating Widgets') is shown when users open the link to that page.

File: animating-widgets.md (hunk 2 above)
Comment (qchong) [inline animating-widgets.md]: changed to "library".

File: animating-widgets.md (hunk 4 above)
Comment (qchong) [inline animating-widgets.md]: fixed.

File: animating-widgets.md (hunk 3 above)
Comment (qchong) [inline animating-widgets.md]: fixed.

File: animations.md (hunk 20 above)
Comment (qchong) [inline animations.md]: This is intended markdown formatting to keep the new paragraph part of the same indent block as its antecedent.

File: animations.md (hunk 27 above)
Comment (qchong) [inline animations.md]: fixed.

File: animations.md (hunk 26 above)
Comment (qchong) [inline animations.md]: fixed

File: animations.md (hunk 28 above)
Comment (qchong) [inline animations.md]: fixed.

**Comment by qchong** [comment]
This PR has a dependency on https://github.com/flutter/flutter/issues/1322 for improved animation code samples.

**Comment by eseidelGoogle** [comment]
@sethladd maybe you can find a new owner for this?

**Comment by sethladd** [comment]
Closing as stale. We can reopen if a new owner is found.
